File: services/nuldoc/nuldoc-src/pages/NotFoundPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import AboutGlobalHeader from "../components/AboutGlobalHeader.ts";
import BlogGlobalHeader from "../components/BlogGlobalHeader.ts";
import SlidesGlobalHeader from "../components/SlidesGlobalHeader.ts";
import DefaultGlobalHeader from "../components/DefaultGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import { Config } from "../config.ts";
import { elem, Element } from "../dom.ts";

export default async function NotFoundPage(
  site: "default" | "about" | "blog" | "slides",
  config: Config,
): Promise<Element> {
  const GlobalHeader = site === "about"
    ? AboutGlobalHeader
    : site === "blog"
    ? BlogGlobalHeader
    : site === "slides"
    ? SlidesGlobalHeader
    : DefaultGlobalHeader;

  return await PageLayout({
    metaCopyrightYear: config.site.copyrightYear,
    metaDescription: "リクエストされたページが見つかりません",
    metaTitle: `Page Not Found｜${config.sites[site].siteName}`,
    site,
    config,
    children: elem(
      "body",
      { class: "single" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        elem("article", {}, elem("div", { class: "not-found" }, "404")),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/not_found_page.rb
module Nuldoc
  module Pages
    class NotFoundPage
      def initialize(site:, config:)
        @site = site
        @config = config
      end

      def render
        site = @site
        config = @config
        global_header = case site
                        when 'about' then Components::AboutGlobalHeader
                        when 'blog' then Components::BlogGlobalHeader
                        when 'slides' then Components::SlidesGlobalHeader
                        else Components::DefaultGlobalHeader
                        end

        site_entry = config.site_entry(site)

        Components::PageLayout.new(
          meta_copyright_year: config.site.copyright_year,
          meta_description: 'リクエストされたページが見つかりません',
          meta_title: "Page Not Found｜#{site_entry.site_name}",
          site: site,
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'single' do
              render(global_header, config: config)
              main class: 'main' do
                article { div(class: 'not-found') { text '404' } }
              end
              render(Components::GlobalFooter, config: config)
            end
          end
        ).build
      end
    end
  end
end
